### ISSUE TITLE
Nested `keyword_list`s do not take default values

### ISSUE THREAD

**Issue by milmazz** [dashbitco/nimble_options#49]
Based on this schema, which, as far as I know, at the root is a `keyword_list`:

```elixir
iex(3)> schema = [
...(3)>   max_connections: [type: :pos_integer, default: 100],
...(3)>   timeout: [type: :pos_integer, default: :timer.seconds(5)]
...(3)> ]
[
  max_connections: [type: :pos_integer, default: 100],
  timeout: [type: :pos_integer, default: 5000]
]
iex(4)> NimbleOptions.validate([], schema)
{:ok, [timeout: 5000, max_connections: 100]}
```

Works as expected, meaning that if you don't provide any value for those options, it will take the default value. But when you have a nested `keyword_list` like this:

```elixir
iex(8)> schema = [
...(8)>   http_adapter: [
...(8)>     type: :keyword_list,
...(8)>     keys: [
...(8)>       max_connections: [type: :pos_integer, default: 100],
...(8)>       timeout: [type: :pos_integer, default: 5000]
...(8)>     ]
...(8)>   ]
...(8)> ]
[
  http_adapter: [
    type: :keyword_list,
    keys: [
      max_connections: [type: :pos_integer, default: 100],
      timeout: [type: :pos_integer, default: 5000]
    ]
  ]
]
iex(9)> NimbleOptions.validate([], schema)
{:ok, []}
```

You don't get the same behavior, I mean, the response from `NimbleOptions.validate([], schema)` should be `{:ok, [http_adapter: [timeout: 5000, max_connections: 100]]}`, right?

But, when you "execute" at least one key from the nested `keyword_list` it works as expected:

```elixir
iex(10)> NimbleOptions.validate([http_adapter: [max_connections: 50]], schema)
{:ok, [http_adapter: [timeout: 5000, max_connections: 50]]}
```

Shouldn't we traverse the nested `keyword_list` regardless if it's given or not as key looking for default values inside?

As a _workaround_, or I don't know if this is the recommended schema configuration, you can define the following:

```elixir
iex(11)> schema = [
...(11)>   http_adapter: [
...(11)>     type: :keyword_list,
...(11)>     default: [max_connections: 100, timeout: :timer.seconds(5)],
...(11)>     keys: [
...(11)>       max_connections: [type: :pos_integer, default: 100],
...(11)>       timeout: [type: :pos_integer, default: :timer.seconds(5)]
...(11)>     ]
...(11)>   ]
...(11)> ]
[
  http_adapter: [
    type: :keyword_list,
    default: [max_connections: 100, timeout: 5000],
    keys: [
      max_connections: [type: :pos_integer, default: 100],
      timeout: [type: :pos_integer, default: 5000]
    ]
  ]
]
iex(12)> NimbleOptions.validate([], schema)
{:ok, [http_adapter: [max_connections: 100, timeout: 5000]]}
iex(13)> NimbleOptions.validate([http_adapter: [timeout: 2000]], schema)
{:ok, [http_adapter: [max_connections: 100, timeout: 2000]]}
iex(14)> NimbleOptions.validate([http_adapter: []], schema)
{:ok, [http_adapter: [timeout: 5000, max_connections: 100]]}
```

But it feels kind of redundant having this block:

```elixir
    default: [max_connections: 100, timeout: 5000],
    keys: [
      max_connections: [type: :pos_integer, default: 100],
      timeout: [type: :pos_integer, default: 5000]
    ]
```

Don't you think? Please let me know what do you think about this.

**Comment by msaraiva** [comment]
Hi @milmazz!

> Shouldn't we traverse the nested `keyword_list` regardless if it's given or not as key looking for default values inside?

I think we shouldn't traverse them and set the default values automatically because we don't know if the user actually wants that parent option to be set or not. If the option was not set and there's no default value, it means it shouldn't be set. In your case, `http_adapter` should be set but other options in different scenarios may not. 

However, by setting the default value to `[]` should indicate that you want the value to always be there so in this case, it should handle the nested default values properly. This PR  #50 addresses this issue so you will be able to do:

```elixir
[
  http_adapter: [
    type: :keyword_list,
    default: [],
    keys: [
      max_connections: [type: :pos_integer, default: 100],
      timeout: [type: :pos_integer, default: 5000]
    ]
  ]
]
```

**Comment by msaraiva** [comment]
Fixed in #50.

**Comment by milmazz** [comment]
@msaraiva Thanks, that helps to avoid the data duplication 🥇
